### DISCUSSION TITLE
Current state of MIR vs QBE?

### DISCUSSION THREAD

**Discussion by rempas** [vnmakarov/mir#273] [General]
First of all, sorry if this is annoying! In my eyes, this is a classic question. It's like "GCC vs Clang", "Linux vs Windows", "C++ vs D", "Barcelona vs Real Madrid" etc. and it will never stop been asked (and for a good reason). I think that QBE is the only other IR that can compete with MIR (at least from the ones I know and excluding "Vox" as it doesn't have some things that I need) so this question is very important and it has to be reviewed every couple of months. Especially now, it doesn't help that there are no benchmarks against QBE but only against other C compilers. So I really believe that we need official benchmarks against QBE to be taken at least once!

Now, I want to make 2 things clear!
1. I'm not just asking out of curiosity (not that this would be bad of course) but because I am searching for the best possible backend for my language NOW. Not in the future, I'm building the compiler and based on my calculations, I'll be able to create something small in a few weeks (and I would have already made it if I wasn't very obsessed with optimize it from the start!) so I need to pick one and I don't have much time left...
2. I have tried to compare them myself but I wasn't able to compile the source code of "tcc" using "c2m". Of course, "c2m" is not part of the "libmir" but still, I thought about making comparisons in big projects that are written in C so I thought about using "c2m" and "cproc" but unfortunately, this didn't went as planned...

Now that I'm clear, the thing is pretty clear. How does MIR compares with QBE? I'm talking about compilation times, runtime performance of the output, features of the library, ease of use and in general, any comparison that can be made between them.

For what I can see myself, QBE (cproc) seems to truly compile very fast and it indeed seems to have at least 70% of "GCC -O2" runtime performance as it claims, it's IR also seems pretty easy and straightforward. Both projects seem to not be very active (at least based on how often commits are uploaded), MIR has more backends (wow!!!). And.... this is all I know!

Thank you for your time regardless! Keep up the amazing work!

**Comment by vnmakarov** [comment]
> Now that I'm clear, the thing is pretty clear. How does MIR compares with QBE? I'm talking about compilation times, runtime performance of the output, features of the library, ease of use and in general, any comparison that can be made between them.
> 
There is a performance comparison of many C compilers on small benchmarks in my blogpost 
https://developers.redhat.com/blog/2021/04/27/the-mir-c-interpreter-and-just-in-time-jit-compiler#how_the_mir_c_compiler_compares_with_other_c_compilers

Basically cproc with QBE generates 20% slower code than `c2m` with MIR.  And `c2m` compilation speed is 1.5 times faster than cproc one.  I have no numbers of speed of MIR vs QBE itself though.

You can run these benchmarks by yourself if you have cproc in your PATH

https://github.com/vnmakarov/mir/blob/761fc31cb1f96d828b90d09c4b7fdbaf4c8440e1/c-benchmarks/run-benchmarks.sh#L93
 
> For what I can see myself, QBE (cproc) seems to truly compile very fast and it indeed seems to have at least 70% of "GCC -O2" runtime performance as it claims, it's IR also seems pretty easy and straightforward. Both projects seem to not be very active (at least based on how often commits are uploaded), MIR has more backends (wow!!!). And.... this is all I know!

QBE does not support long double as I know.  I am not sure QBE is ABI compatible even besides long double part.

I am still actively working on MIR.  GITHUB shows changes only for master branch therefore MIR looks like stalled.  But I am doing most my work on bbv branch.  This branch will be the base for the next MIR release and I am planning to make at least one release this year.  The branch is not release ready but I am using it for my work on MIR-based JIT for CRuby.



> 
> Thank you for your time regardless! Keep up the amazing work!

Thank you for your kind words.

**Reply by rempas** [reply]
So you're telling me that MIR has both 20% (20!!!!!) better runtime and 50% faster compilation times than QBE? How I wasn't aware of that until now? Thank you for uploading the benchmark but I'm trusting you (because why lying anyway?) so I'll pass running it to save some time. Either way, the results will vary between CPU (with older ones having bigger differences). Also, `c2m` is written by you (right?) so it makes the best usage of `libmir` and given the fact of how simple QBE's IR is, I would expect that `cproc` also does the best usage as well. In the worse scenario, probably there should be a loss of 5% so still the difference is really big!

I'm also really happy that not only you answered but you did is so fast (while I didn't got an answer when I made a question in QBE's mailing list)! Showing the person behind the code is always nice to me ;)

This motivates me to work harder and have it ready as fast as possible! Tbh, finding a backend was always the problem and what give me the vibes to abandon the project at all... Of course, hand written assembly or even better fully hand written machine language and ELF files were a thing but let's be honest! Not only I hate reading (and this things will require a lot of reading) but, even If I get discipline, it will still take a lot of years to learn and implement (and make sure they work perfect) all these stuff and well... why not use the work someone else has already done? Isn't this why we have open-source code in the beginning? People are writing code and they expect other people to have fun and use it!

That's also my motivation to make my language! I'm fascinated by the open source community and what they have made! So I want to offer something that people will love and want to use (to create even more awesome stuff!). So someone has already spent this thousands hours, so use their work! And of course the faster I introduce the language, the better! I'll try to release a version this year as well! Fingers crossed!

Should I use `master` or `bbv`? Of course, `bbv` is not consider a "stable release" channel but still, every commit is tested (right?) so I suppose it is safe. Other than that, are there any other reasons to not consider using it?

**Reply by vnmakarov** [reply]
> Should I use `master` or `bbv`? Of course, `bbv` is not consider a "stable release" channel but still, every commit is tested (right?) so I suppose it is safe. Other than that, are there any other reasons to not consider using it?

Sorry for delay with the answer.  I recommend to use master.  bbv branch is not stable and has some known bugs of new features (e.g. in register variables and anti-alias attribute usage) and how to solve the bugs still needs some thinking for me.  You can switch to bbv later if you need really new features and probably a better generated code.

**Reply by rempas** [reply]
Thank you! Good luck and have fun!

**Reply by rempas** [reply]
> You can switch to bbv later if you need really new features and probably a better generated code.

@vnmakarov Hello! I was thinking about some stuff and while I was working, the sentence I quoted crossed my mind.
What do you mean "better generated code"? Better in what?

**Reply by vnmakarov** [reply]
> What do you mean "better generated code"? Better in what?
It means performance.  The branch contains optimizations which help to improve JITted code for stack based VMs and specifically CRuby VM.

For example, the branch contains optimizations which remove redundant loads/stores.  The branch implements register variables (variables which use fixed hard regs) and C extension to describe this.  It also implements calls and returns by jumps and C builtins for this (which is important for direct switching to direct threaded interpreter).  builtin_expect is also implemented on the branch to generate better code locality.  The branch also implements anti-alias attributes to differ pointers whose values are never intersected (but it is buggy right now).

**Reply by rempas** [reply]
Wow! I just don't know what to say! You're doing MORE optimizations?! And again, the super crazy compile times! I'm really speechless! I did tried to learn how machine instruction syntax works for the fun of it and it is hell!!! (at least for x86-64) I cannot imagine that you are doing this and you are even doing optimizations that can get close to the BEAST called GCC. Your little gem deserves a great language frontend and I'm working as hard as I can to make one!

And how knows, maybe you'll want to re-write MIR on in if you end up liking it! :rofl:
